src/tss2-tcti/tcti-tbs.c

```diff
@@ -0,0 +1,286 @@
+ */
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <windows.h>
+#include <Tbs.h>
+
+#include "tss2_tcti.h"
+#include "tss2_tcti_tbs.h"
+
+#include "tcti-common.h"
+#include "tcti-tbs.h"
+#define LOGMODULE tcti
+#include "util/log.h"
+
+/*
+ * This function wraps the "up-cast" of the opaque TCTI context type to the
+ * type for the TBS TCTI context. The only safe-guard we have to ensure
+ * this operation is possible is the magic number for the TBS TCTI context.
+ * If passed a NULL context, or the magic number check fails, this function
+ * will return NULL.
+ */
+TSS2_TCTI_TBS_CONTEXT*
+tcti_tbs_context_cast (TSS2_TCTI_CONTEXT *tcti_ctx)
+{
+    if (tcti_ctx != NULL && TSS2_TCTI_MAGIC (tcti_ctx) == TCTI_TBS_MAGIC) {
+        return (TSS2_TCTI_TBS_CONTEXT*)tcti_ctx;
+    }
+    return NULL;
+}
+/*
+ * This function down-casts the TBS TCTI context to the common context
+ * defined in the tcti-common module.
+ */
+TSS2_TCTI_COMMON_CONTEXT*
+tcti_tbs_down_cast (TSS2_TCTI_TBS_CONTEXT *tcti_tbs)
+{
+    if (tcti_tbs == NULL) {
+        return NULL;
+    }
+    return &tcti_tbs->common;
+}
+
+TSS2_RC
+tcti_tbs_transmit (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t command_size,
+    const uint8_t *command_buffer)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_tbs_down_cast (tcti_tbs);
+    TSS2_RC rc = TSS2_RC_SUCCESS;
+    TBS_RESULT tbs_rc;
+    /* Reset resSize */
+    tcti_tbs->resSize = TBS_RESULT_MAX_BUFFER_SIZE;
+
+    if (tcti_tbs == NULL) {
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+    rc = tcti_common_transmit_checks (tcti_common, command_buffer);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+    LOGBLOB_DEBUG (command_buffer,
+                   command_size,
+                   "sending %zu byte command buffer:",
+                   command_size);
+    tbs_rc = Tbsip_Submit_Command (tcti_tbs->hContext,
+                                   TBS_COMMAND_LOCALITY_ZERO,
+                                   TBS_COMMAND_PRIORITY_NORMAL,
+                                   command_buffer,
+                                   command_size,
+                                   tcti_tbs->resultBuffer,
+                                   &(tcti_tbs->resSize)
+    );
+    if (tbs_rc != TBS_SUCCESS) {
+        LOG_ERROR ("Failed to submit command to TBS with error: 0x%x", tbs_rc);
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    tcti_common->state = TCTI_STATE_RECEIVE;
+    return TSS2_RC_SUCCESS;
+}
+
+TSS2_RC
+tcti_tbs_receive (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t *response_size,
+    uint8_t *response_buffer,
+    int32_t timeout)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_tbs_down_cast (tcti_tbs);
+    TSS2_RC rc = TSS2_RC_SUCCESS;
+
+    if (tcti_tbs == NULL) {
```

Comment:
Same here, could this be brought up? Do y'all stick to the C89 standard of having the variable declarations all at the very top?

Reply:
Yeah, this repo generally sticks to having all variable declarations at the top, so I think it would be best to keep this the way it is to keep things consistent.